text/0063-coprocessor-plugin.md

```diff
@@ -0,0 +1,323 @@
+pub trait CoprocessorPlugin: Send + Sync {
+    /// Handles a request to the coprocessor.
+    ///
+    /// The data in the `request` parameter is exactly the same data that was passed with the
+    /// `RawCoprocessorRequest` in the `data` field. Each plugin is responsible to properly decode
+    /// the raw bytes by itself.
+    /// The same is true for the return parameter of this function. Upon successful completion, the
+    /// function should return a properly encoded result as raw bytes which is then sent back to
+    /// the client.
+    ///
+    /// Most of the time, it's a good idea to use Protobuf for encoding/decoding, but in general you
+    /// can also send raw bytes.
+    ///
+    /// Plugins can read and write data from the underlying [`RawStorage`] via the `storage`
+    /// parameter.
+    fn on_raw_coprocessor_request(
+        &self,
+        ranges: Vec<Range<Key>>,
+        request: RawRequest,
+        storage: &dyn RawStorage,
+    ) -> PluginResult<RawResponse>;
+}
+
+/// Storage access for coprocessor plugins.
+///
+/// [`RawStorage`] allows coprocessor plugins to interact with TiKV storage on a low level.
+///
+/// Batch operations should be preferred due to their better performance.
+#[async_trait(?Send)]
+pub trait RawStorage {
+    /// Retrieves the value for a given key from the storage on the current node.
+    /// Returns [`Option::None`] if the key is not present in the database.
+    async fn get(&self, key: Key) -> PluginResult<Option<Value>>;
+
+    /// Same as [`RawStorage::get()`], but retrieves values for multiple keys at once.
+    async fn batch_get(&self, keys: Vec<Key>) -> PluginResult<Vec<KvPair>>;
+
+    /// Same as [`RawStorage::get()`], but accepts a `key_range` such that values for keys in
+    /// `[key_range.start, key_range.end)` are retrieved.
+    /// The upper bound of the `key_range` is exclusive.
+    async fn scan(&self, key_range: Range<Key>) -> PluginResult<Vec<Value>>;
+
+    /// Inserts a new key-value pair into the storage on the current node.
+    async fn put(&self, key: Key, value: Value) -> PluginResult<()>;
+
+    /// Same as [`RawStorage::put()`], but inserts multiple key-value pairs at once.
+    async fn batch_put(&self, kv_pairs: Vec<KvPair>) -> PluginResult<()>;
+
+    /// Deletes a key-value pair from the storage on the current node given a `key`.
+    /// Returns [`Result::Ok]` if the key was successfully deleted.
+    async fn delete(&self, key: Key) -> PluginResult<()>;
+
+    /// Same as [`RawStorage::delete()`], but deletes multiple key-value pairs at once.
+    async fn batch_delete(&self, keys: Vec<Key>) -> PluginResult<()>;
+
+    /// Same as [`RawStorage::delete()`], but deletes multiple key-values pairs at once
+    /// given a `key_range`. All records with keys in `[key_range.start, key_range.end)`
+    /// will be deleted. The upper bound of the `key_range` is exclusive.
+    async fn delete_range(&self, key_range: Range<Key>) -> PluginResult<()>;
```

Comment:
how about separating the supporting of the coprocessor plugin into 2 phases:
- the first phase is to support read-only APIs, enable TiKV users to add custom key-value codecs, data types (models), and calculate function and models.
- the second phase is to support write-related APIs, enable tiKV users to pushdown custom write operations, improve the write performance